### ISSUE TITLE
Improve handling for LHE files.

### ISSUE THREAD

**Issue by annawoodard** [matz-e/lobster#108]
We are currently unable adjust job size for LHE files (we are restricted to 1 file / jobit.) We should add 'skip events' to our jobit definition (file, run, lumi) --> (file, run, lumi, skip events) which will enable us to split one LHE file into multiple jobs, along with an LHE interface to count the events per file.

I've started a draft of this out of necessity, so I'm going to assign it to myself.

**Comment by matz-e** [comment]
Could we abuse run, lumi instead?  Basically have two free parameters for LHE:

(file, run, lumi) for data processing --> (file, start_event, end_event) for LHE

That way we would avoid having tons of columns in the jobit database, which will increase storage needs.

**Comment by annawoodard** [comment]
Hmm. Now that I think about it, for the LHE case we only need (file, skip_events) (I'm insisting on using skip_events because that is consistent with the pool source definition). The third parameter is 'max events', and it only needs to be defined once per dataset.

**Comment by matz-e** [comment]
But per job you need an amount of events to be processed?  It may be fixed for the dataset, but could still translate into a third parameter.

Also if you're running with more than one jobit per job, you are going to run into trouble, though, as LHE does not provide any kind of masking.

An idea I have had earlier is to drop the run, lumi columns and introduce a more advanced "id" column.  For data, this could just save a (run, lumi) tuple, for your LHE needs the simple "skip_events" is sufficient.

**Comment by matz-e** [comment]
And this generic column could also include the argument column you introduced a while back…

**Comment by annawoodard** [comment]
> But per job you need an amount of events to be processed?  It may be fixed for the dataset, but could still translate into a third parameter.

We would have ```max_events / job = max_events / jobit * jobits / job```, where ```max_events / jobit``` and ```jobits / job``` could be stored once in the datasets table rather than for each jobit.

> Also if you're running with more than one jobit per job, you are going to run into trouble, though, as LHE does not provide any kind of masking.

We could do multiple jobits / job as long as we requested them in order from the DB and ```job_skip_events = min(jobit_skip_events)``` and job_max_events as defined above. If we wanted to trade simplicity for flexibility, we could have the jobsize fixed. 

> An idea I have had earlier is to drop the run, lumi columns and introduce a more advanced "id" column.  For data, this could just save a (run, lumi) tuple, for your LHE needs the simple "skip_events" is sufficient.

Would you prefer to add an additional column ("units" maybe) to the dataset table to specify the type of jobit id (so we know how to handle it)? Or alternatively, try to duck type? If we did add that column, I think we could remove the 'empty_source' column.

**Comment by matz-e** [comment]
> But per job you need an amount of events to be processed? It may be fixed for the dataset, but could still translate into a third parameter.
> 
> We would have max_events / job = max_events / jobit * jobits / job, where max_events / jobit and jobits / job could be stored once in the datasets table rather than for each jobit.
> 
jobits / job is already in the dataset table, indicating the task size.  I think that the max_events / jobit should go into the jobits table in one form or the other.  I would rather not add too many columns to the dataset table, but try to keep some flexibility.  Otherwise, all the special cases might be hard to disentangle and add to later on.
> Also if you're running with more than one jobit per job, you are going to run into trouble, though, as LHE does not provide any kind of masking.
> 
> We could do multiple jobits / job as long as we requested them in order from the DB and job_skip_events = min(jobit_skip_events) and job_max_events as defined above. If we wanted to trade simplicity for flexibility, we could have the job size fixed.
> 
Keeping the job size fixed would be the easiest way, I think.  Unless job sizes are way out of league, ad-hoc adjustments shouldn’t be necessary.
> An idea I have had earlier is to drop the run, lumi columns and introduce a more advanced "id" column. For data, this could just save a (run, lumi) tuple, for your LHE needs the simple "skip_events" is sufficient.
> 
> Would you prefer to add an additional column ("units" maybe) to the dataset table to specify the type of jobit id (so we know how to handle it)? Or alternatively, try to duck type? If we did add that column, I think we could remove the 'empty_source' column.
> 
Yes, a units / mode column would be good (esp. if we can drop another column.)  In my mind, this should also make it easier in the future to add other processing modes.  All someone has to provide is a way to specify this unique “id” column and translate it into a job configuration.

**Comment by annawoodard** [comment]
> jobits / job is already in the dataset table, indicating the task size. 

Yes, of course, sorry I wasn't clear before-- I meant that we could add the additional column ```max_events / jobit``` (not meaning to imply that the jobsize wasn't already there).

> I think that the max_events / jobit should go into the jobits table in one form or the other.  

I strongly disagree. Why add a column for a number that is going to be the same for every single row? That's the reason for having a datasets table.

> Keeping the job size fixed would be the easiest way, I think.  Unless job sizes are way out of league, ad-hoc adjustments shouldn’t be necessary.

I weakly disagree. Having adjustable jobsizes is one of Lobster's best features. We want to be able to plug different job-creation modes in, but have the higher-level behavior be transparent to various job-creation modes. Having some modes where the jobsize is adjustable and some where it isn't complicates the higher-level behavior in favor of simpler lower-level behavior, but I think it should be the other way around.

> Yes, a units / mode column would be good (esp. if we can drop another column.)  In my mind, this should also make it easier in the future to add other processing modes.  All someone has to provide is a way to specify this unique “id” column and translate it into a job configuration.

Sounds good, I'll go with that, then.

**Comment by matz-e** [comment]
>    I strongly disagree. Why add a column for a number that is going to be the same for every single row? That's the reason for having a datasets table.

But for running over collisions, this can be estimated by `events in file / lumis in file`, which may not be constant over a dataset!  I don't think we should add specialized columns everywhere, as that will make changing things in the future a lot harder.

> I weakly disagree. Having adjustable jobsizes is one of Lobster's best features. We want to be able to plug different job-creation modes in, but have the higher-level behavior be transparent to various job-creation modes. Having some modes where the jobsize is adjustable and some where it isn't complicates the higher-level behavior in favor of simpler lower-level behavior, but I think it should be the other way around.

OK.  I just think that with our current model of drawing random jobits and building tasks out of them, treating the LHE files is such a special case that it will complicate things a lot.  A fixed jobsize doesn't complicate things IMO, as it is a specialization of the adjustable size.

**Comment by annawoodard** [comment]
> But for running over collisions, this can be estimated by `events in file / lumis in file`, which may not be constant over a dataset!  I don't think we should add specialized columns everywhere, as that will make changing things in the future a lot harder.

Ok, but I'll just point out that we don't currently use the quantity `events in file / lumis in file` for anything (we use `events in dataset / lumis in dataset` to estimate the jobsize once per dataset-- we probably should do it per file, I've opened #110). Anyways, that quantity is *not* equivalent to max_events, because max_events should only be used to set the maximum number of events to process (if our estimate is wrong, we miss events.) So it's not a 'specialized column'  in the sense that it has a different meaning for one mode over another: it always means 'max events to process per job' (-1 for collisions, or given by the user for MC production jobs).

> OK.  I just think that with our current model of drawing random jobits and building tasks out of them, treating the LHE files is such a special case that it will complicate things a lot.  A fixed jobsize doesn't complicate things IMO, as it is a specialization of the adjustable size.

I agree it will add a little complexity (and I am usually fighting in favor of simplicity at the expense of extra functionality, what's going on here?), but I think that dynamic job sizing is so central to the Lobster spirit that we should preserve it for all running modes. This will also be relevant if we manage to implement #74: in that case, having a fixed jobsize will mean that we do need to fundamentally treat production jobs differently (not just have them as a specialization of processing jobs.)

**Comment by matz-e** [comment]
> Ok, but I'll just point out that we don't currently use the quantity events in file / lumis in file for anything (we use events in dataset / lumis in dataset to estimate the jobsize once per dataset-- we probably should do it per file, I've opened #110). Anyways, that quantity is not equivalent to max_events, because max_events should only be used to set the maximum number of events to process (if our estimate is wrong, we miss events.) So it's not a 'specialized column' in the sense that it has a different meaning for one mode over another: it always means 'max events to process per job' (-1 for collisions, or given by the user for MC production jobs).

We should use `events in file / lumis in file` per jobit, then the job creation is a lot easier, as otherwise you'd have to cross-query the files the whole time.  Coming up with an abstraction to merge lumi-, file-, and job-based processing is then not so hard.  LHE files would still be a special case, due to the job-size behavior.

I see your point about `max events per job`, but I still think it would be worth-while to save it in the jobits table, as then your jobits are fully defined in just one table.

> I agree it will add a little complexity (and I am usually fighting in favor of simplicity at the expense of extra functionality, what's going on here?), but I think that dynamic job sizing is so central to the Lobster spirit that we should preserve it for all running modes. This will also be relevant if we manage to implement #74: in that case, having a fixed jobsize will mean that we do need to fundamentally treat production jobs differently (not just have them as a specialization of processing jobs.)

It's still one very special treatment for LHE files.  Again, I agree that we want to eventually have this, and right now I feel like it will result in an additional job creation method, otherwise it'll complicate our current one a whole lot more than I would like to see for #65 , as I see it.

Oh, why does LHE have to be the odd one out?

**Comment by annawoodard** [comment]
> We should use `events in file / lumis in file` per jobit, then the job creation is a lot easier, as otherwise you'd have to cross-query the files the whole time.  Coming up with an abstraction to merge lumi-, file-, and job-based processing is then not so hard.  LHE files would still be a special case, due to the job-size behavior.

Agreed.

> I see your point about `max events per job`, but I still think it would be worth-while to save it in the jobits table, as then your jobits are fully defined in just one table.

Agreed.

> Again, I agree that we want to eventually have this, and right now I feel like it will result in an additional job creation method, otherwise it'll complicate our current one a whole lot more than I would like to see for #65

I think that LHE handling should be higher priority than #65, not the other way around. The first-order results of implementing #65 will be:
1. More elegant code
2. Easier-to-implement future features
3. Get us most of the way to implementing #110-- but note that is an *improvement* on an existing feature, and we don't have hard evidence that variation in lumi size is really an important problem for us now. On the other hand, we *do* know that not being able to adjust the jobsize, if it is chosen incorrectly, means that the entire project will need to be restarted-- this is a fundamental problem that we set out to solve when we started Lobster.

So while #65 is is very important, the main result is improvements, while the main result of implementing dynamic jobsizing for LHE files would be to preserve a currently-existing feature which is broken for an important production mode that we need in order to do physics.

**Comment by matz-e** [comment]
I agree that we'll probably see more benefit from prioritizing this!
